### PR TITLE
Adaptive sampling

### PR DESCRIPTION
**Pull request by Jeyhun1**
Added adaptive sampling to the framework and applied it in 1D Heat Equation and 1D Burgers Equation examples

### CODE REVIEW

File: PINNFramework/PDELoss.py [hunk 1]
```diff
@@ -31,3 +31,33 @@ def __call__(self, x: Tensor, model: Module, **kwargs):
         pde_residual = self.pde(x, u, **kwargs)
         zeros = torch.zeros(pde_residual.shape, device=pde_residual.device)
         return self.norm(pde_residual, zeros)
+
+
+class PDELossAdaptive(PDELoss):
```
Comment (StillerPatrick) [inline PINNFramework/PDELoss.py:36]: Why do we need an additional loss term? That's make the framework too complicated. We need a flexible implementation of PDE loss, which can handle multiple sampler

**Comment by StillerPatrick** [comment]
My remarks:
- schrödinger equation is not in geometry fashion
- PDE loss should allocate the next points from the sampler 
- adaptive PDE loss is ok
- Geometry should be a module such that I have PINNFramework.Geometry
- Geometry should be changed to an interface such that we can implement  multiple ones it
- Geometry should have sampler insight
- Sampler should be a module such that I have PINNFramework.Sampler
- we need some standard samplers such as lhs, random etc. 
- sampler should be implemented from an interface that allows implementing own sampling methods in the future.

File: PINNFramework/Geometry.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+class Geometry(ABC):
+    def __init__(self, lb, ub, n_points, sampler= 'LHS', device = torch.device("cuda")):
+        """
+        Constructor of the Geometry class
+
+        Args:
+            lb (numpy.ndarray): lower bound of the domain.
+            ub (numpy.ndarray): upper bound of the domain.
+            n_points (int): the number of sampled points.
+            sampler (string): sampling method: "random" (pseudorandom), "LHS" (Latin hypercube sampling), and "adaptive" method.
```
Comment (StillerPatrick) [inline PINNFramework/Geometry.py:14]: Please split in two lines.

**Comment by StillerPatrick** [comment]
@sutariyaraj please review the code as well to check if it fits your current implementation of the inference pipeline. Please communicate prober changes.

File: PINNFramework/Sampler.py [hunk 3]
```diff
@@ -2,78 +2,117 @@
...
-
-        # obtain 'num_points' indices sampled from the multinomial distribution
-        indicies_new = torch.multinomial(q_model[:, 0], self.num_points, replacement=True)
-        
-        # collocation points and corresponding weights
-        self.xf = xf[indicies_new]
-        self.weight = q_model[indicies_new]
-
-      return self.xf, self.weight
+seed = 42
```
Comment (nih23) [inline PINNFramework/Sampler.py:7]: hard coded magic number

File: PINNFramework/Sampler.py [hunk 4]
```diff
@@ -2,78 +2,117 @@
...
 
 
+def sample(lb, ub, num_points, sampler, num_seed, model, pde, device = torch.device("cuda:0")):
```
Comment (nih23) [inline PINNFramework/Sampler.py]: the definition of this function is polluted and has to be cleaned because you are asking for parameters that some samples like quasirandym are not using. I suggest to go for an object-oriented rewrite of the full sampling approach.

File: PINNFramework/Sampler.py [hunk 5]
```diff
@@ -2,78 +2,117 @@
...
+    """Pseudo random sampling. Returns sampled points in [lb,ub].
+    Args:
+        lb (numpy.ndarray): lower bound of the domain.
+        ub (numpy.ndarray): upper bound of the domain.
+        num_points (int): the number of sampled points.
+    """
+    
+    np.random.seed(seed)
+    dimension = lb.shape[1]
+    xf = np.random.uniform(lb,ub,size=(num_points, dimension))
```
Comment (nih23) [inline PINNFramework/Sampler.py]: use torch.distribution instead of np.random to avoid unnecessary IO between host (numpy) and device (torch)

File: PINNFramework/Sampler.py [hunk 6]
```diff
@@ -2,78 +2,117 @@
...
+    q_model = torch.softmax(losses_xf, dim=0)
+
+    # obtain 'num_points' indices sampled from the multinomial distribution
+    indicies_new = torch.multinomial(q_model[:, 0], num_points, replacement=True)
+
+    # collocation points and corresponding weights
+    xf = xf[indicies_new]
+    weight = q_model[indicies_new].detach()
+    #print('xf type', type(xf))
+    print('xf ', (xf))
```
Comment (nih23) [inline PINNFramework/Sampler.py]: polluted output

**Review by nih23 — CHANGES_REQUESTED** [review]

File: PINNFramework/Sampler.py (hunk 6 above)
Comment (nih23) [reply PINNFramework/Sampler.py]: the print statements from line 113 result in unnecessary verbose output

File: PINNFramework/Sampler.py (hunk 4 above)
Comment (StillerPatrick) [reply PINNFramework/Sampler.py]: I agree. I think we should go back to the OOP Version of the Sampler (sorry for that). This would make the code better to understand and maintainable.

Comment (StillerPatrick) [reply PINNFramework/Sampler.py]: I also miss a definition of the total points we want to sample. This determines the number of iterations we go through the dataset.

File: PINNFramework/Sampler.py (hunk 5 above)
Comment (StillerPatrick) [reply PINNFramework/Sampler.py]: We definitely check this option as possible performance improvement

File: PINNFramework/Sampler.py (hunk 6 above)
Comment (Jeyhun1) [reply PINNFramework/Sampler.py]: Those print statements are deleted in later commits

**Comment by StillerPatrick** [comment]
Approved. Nice work
